File: Source/Java/Expressive/src/main/java/com/bijington/expressive/operators/OperatorBase.java
package com.bijington.expressive.operators;

import com.bijington.expressive.expressions.IExpression;
import com.sun.xml.internal.ws.policy.privateutil.PolicyUtils;

import java.util.List;

/**
 * Created by shaun on 08/07/2016.
 */
public abstract class OperatorBase implements IOperator {
    public abstract String[] getTags();

    public abstract IExpression buildExpression(String previousToken, IExpression[] expressions);

    public Boolean canGetCaptiveTokens(String previousToken, String token, List<String> remainingTokens)
    {
        return true;
    }

    public String[] getCaptiveTokens(String previousToken, String token, List<String> remainingTokens)
    {
        return new String[] { token };
    }

    public String[] getInnerCaptiveTokens(String[] allCaptiveTokens)
    {
        return new String[0];
    }

    public abstract int getPrecedence(String previousToken);
}

File: Source/Expressive/Operators/OperatorBase.cs
using Expressive.Expressions;
using System.Collections.Generic;

namespace Expressive.Operators
{
    internal abstract class OperatorBase : IOperator
    {
        #region IOperator Members

        public abstract string[] Tags { get; }

        public abstract IExpression BuildExpression(Token previousToken, IExpression[] expressions, Context context);

        public virtual bool CanGetCaptiveTokens(Token previousToken, Token token, Queue<Token> remainingTokens)
        {
            return true;
        }

        public virtual Token[] GetCaptiveTokens(Token previousToken, Token token, Queue<Token> remainingTokens)
        {
            return new[] { token };
        }

        public virtual Token[] GetInnerCaptiveTokens(Token[] allCaptiveTokens)
        {
#pragma warning disable CA1825 // Avoid zero-length array allocations. - Array.Empty does not exist in net 4.5
            return new Token[0];
#pragma warning restore CA1825 // Avoid zero-length array allocations.
        }

        public abstract OperatorPrecedence GetPrecedence(Token previousToken);

        #endregion
    }
}
